Add support for STS duration option

thehunt33r opened the pull request:
This PR adds the possibility to specify a specific STS Session duration when assuming a role via Terragrunt.

Resolves #495

Test output: https://gist.github.com/thehunt33r/895f31f880142e9587dd6afd5937dc7c

From what I gather on the tests, it's only GCP/GCS stuff which fails (due to me not having a GCP account), which I haven't touched

brikis98 commented on aws_helper/config.go:
```diff
@@ -124,7 +124,7 @@ func CreateAwsSession(config *AwsSessionConfig, terragruntOptions *options.Terra
 
 // Make API calls to AWS to assume the IAM role specified and return the temporary AWS credentials to use that role
-func AssumeIamRole(iamRoleArn string) (*sts.Credentials, error) {
+func AssumeIamRole(iamRoleArn string, duration int64) (*sts.Credentials, error) {
```
NIT: Could you use a name that makes the units clearer? E.g., `durationSeconds`?

thehunt33r replied:
Renamed it to `sessionDurationSeconds` which is clearer in terms of what the duration is for, as well as the unit. Thanks for the feedback.

brikis98 commented on aws_helper/config.go:
```diff
@@ -202,8 +203,8 @@ func AssumeRoleAndUpdateEnvIfNecessary(terragruntOptions *options.TerragruntOpti
 
-	terragruntOptions.Logger.Debugf("Assuming IAM role %s", terragruntOptions.IamRole)
-	creds, err := AssumeIamRole(terragruntOptions.IamRole)
+	terragruntOptions.Logger.Debugf("Assuming IAM role %s with duration %d", terragruntOptions.IamRole, terragruntOptions.StsDuration)
```
```suggestion
	terragruntOptions.Logger.Debugf("Assuming IAM role %s with duration %d seconds", terragruntOptions.IamRole, terragruntOptions.StsDuration)
```

brikis98 commented on cli/args.go:
```diff
@@ -110,6 +110,12 @@ func parseTerragruntOptionsFromArgs(terragruntVersion string, args []string, wri
 	}
 
+	envValue, envProvided := os.LookupEnv("TERRAGRUNT_STS_DURATION")
+	stsDuration, err := parseIntArg(args, OPT_TERRAGRUNT_STS_DURATION, envValue, envProvided, 3600)
```
NIT: extract `3600` into a constant with a clear name (e.g., `default_iam_role_duration`) and comment (e.g., `# one hour`).

brikis98 commented on cli/cli_app.go:
```diff
@@ -217,6 +219,7 @@ GLOBAL OPTIONS:
    terragrunt-source                            Download Terraform configurations from the specified source into a temporary folder, and run Terraform in that temporary folder.
    terragrunt-source-update                     Delete the contents of the temporary folder to clear out any old, cached source code before downloading new source code into it.
    terragrunt-iam-role                          Assume the specified IAM role before executing Terraform. Can also be set via the TERRAGRUNT_IAM_ROLE environment variable.
+   terragrunt-sts-duration                      Session duration for STS session. Can also be set via the TERRAGRUNT_STS_DURATION environment variable.
```
NIT: I'm not sure "STS" is going to be as recognizable to many users as "IAM Role." Perhaps `--terragrunt-iam-role-duration` would be a clearer name? This applies to the CLI arg, env var, etc.

thehunt33r replied:
I agree with the fact that `terragrunt-sts-duration` might be a bit short, however `terragrunt-iam-role-duration` seems "wrong" (i.e it's not really the duration of the IAM role, but of the AssumeRole session)

Do you think `terragrunt-iam-assume-role-duration` / `iam_assume_role_duration` / `TERRAGRUNT_IAM_ASSUME_ROLE_DURATION` might be better?

brikis98 replied:
That works for me!

brikis98 commented on cli/cli_app.go:
```diff
@@ -370,6 +373,11 @@ func RunTerragrunt(terragruntOptions *options.TerragruntOptions) error {
 	}
 
+	// replace default sts duration if set in config
+	if terragruntOptions.StsDuration == 3600 && terragruntConfig.StsDuration != nil {
```
Why are you checking for exactly equality to 3600? Do you mean `!=`? Any reason not to always set the param, as long as it's not `nil`?

thehunt33r replied:
The intent here was to check whether `StsDuration` had been overwritten (different from the default of 3600) in `terragruntOptions` (i.e. via CLI or ENV arguments). If it hasn't been overwritten, and `StsDuration` has been defined in `terragruntConfig`, then we should use the `StsDuration` defined in `terragruntConfig`.

I think this is a similar approach to the `DownloadDir` assignment on line 395:397.

brikis98 replied:
Ah, OK. Extracting the 3600 into a constant will help make that clear.

brikis98 commented on config/config_as_cty.go:
```diff
@@ -79,6 +79,15 @@ func terragruntConfigAsCty(config *TerragruntConfig) (cty.Value, error) {
+	if err != nil {
+		return cty.NilVal, err
+	}
+	fmt.Println(stsDurationCty)
```
Left over debugging?

thehunt33r replied:
😅  Oops sorry. Fixed.

brikis98 commented on docs/_docs/04_reference/cli-options.md:
```diff
@@ -381,28 +381,44 @@ This should allow you to run `import` on the module and work around those Terraf
-- [terragrunt-check](#terragrunt-check)
-- [terragrunt-hclfmt-file](#terragrunt-hclfmt-file)
-- [terragrunt-override-attr](#terragrunt-override-attr)
+- [CLI commands](#cli-commands)
```
This is quite a large diff. Was your branch out of date or something?

thehunt33r replied:
Looks like my markdown linter went a bit haywire here. My branch wasn't that out of date (just 1 commit behind HEAD)! I'll fix that :)

brikis98 commented on docs/_docs/04_reference/cli-options.md:
```diff
@@ -525,6 +541,14 @@ When passed in, the `*-all` commands continue processing components even if a de
+**Environment Variable**: `TERRAGRUNT_STS_DURATION`<br/>
+**Requires an argument**: `--terragrunt-sts-duration 3600`
+
+Uses the specified duration as the session duration of the STS credentials assumed for the role defined in `--terragrunt-iam-role`
```
NIT: specify the units for the duration (seconds).

brikis98 commented on docs/_docs/04_reference/config-blocks-and-attributes.md:
```diff
@@ -724,6 +733,19 @@ Example:
 
+### sts_duration
+
+The `sts_duration` attribute can be used to specify an IAM role that Terragrunt should assume prior to invoking Terraform.
```
NIT: specify the units for the duration (seconds).

brikis98 commented on options/options.go:
```diff
@@ -100,6 +100,9 @@ type TerragruntOptions struct {
 	// The ARN of an IAM Role to assume before running Terraform
 	IamRole string
 
+	// Duration of the STS Session
+	StsDuration int64
```
I believe there are `New` and `Clone` methods for `Options` that need to be updated with this param.

thehunt33r replied:
I've updated the Clone and New method and added a default const for the default StsDuration of 3600.

brikis98 reviewed (commented):
Thanks for the PR! And for updating the docs + tests 👍

thehunt33r commented:
Hey @brikis98 

Thank you so much for your review! I've addressed most of your comments.

thehunt33r commented:
Hey @brikis98 , thanks again for your review. 

I've renamed the argument as agreed, as well as extracted the `3600` into a var.

Finally I've re-run the tests and updated the gist. 

Would you like me to clean up the history before we can consider merging?

brikis98 commented on cli/cli_app.go:
```diff
@@ -372,6 +375,12 @@ func RunTerragrunt(terragruntOptions *options.TerragruntOptions) error {
 	}
 
+	// replace default sts duration if set in config
+	defaultIamAssumeRoleDuration := 3600 // Default duration for IAM Assume role session is 3600s (1hr)
+	if terragruntOptions.IamAssumeRoleDuration == int64(defaultIamAssumeRoleDuration) && terragruntConfig.IamAssumeRoleDuration != nil {
```
NIT: remove `defaultIamAssumeRoleDuration` and use `options.DEFAULT_IAM_ASSUME_ROLE_DURATION` here instead.

thehunt33r replied:
Good catch! I hadn't noticed the options package was already imported/used for the purpose of setting consts across the whole project. I've updated this now :)

brikis98 commented on cli/args.go:
```diff
@@ -115,6 +115,13 @@ func parseTerragruntOptionsFromArgs(terragruntVersion string, args []string, wri
 	}
 
+	envValue, envProvided := os.LookupEnv("TERRAGRUNT_IAM_ASSUME_ROLE_DURATION")
+	defaultIamAssumeRoleDuration := 3600 // Default STS sessions are 1 hour long (3600 seconds)
+	IamAssumeRoleDuration, err := parseIntArg(args, OPT_TERRAGRUNT_IAM_ASSUME_ROLE_DURATION, envValue, envProvided, defaultIamAssumeRoleDuration)
```
NIT: remove `defaultIamAssumeRoleDuration` and use `options.DEFAULT_IAM_ASSUME_ROLE_DURATION` here instead.

brikis98 reviewed (commented):
Very close! Just a couple minor tweaks, and I can then kick off tests. Thank you!

thehunt33r commented on config/config.go:
```diff
@@ -85,6 +86,7 @@ type terragruntConfigFile struct {
 	PreventDestroy         *bool               `hcl:"prevent_destroy,attr"`
 	Skip                   *bool               `hcl:"skip,attr"`
 	IamRole                *string             `hcl:"iam_role,attr"`
+	IamAssumeRoleDuration            *int64              `hcl:"iam_assume_role_duration,attr"`
```
oops... sorry my editor was acting up there. Fixed!

thehunt33r commented:
Here we go @brikis98; I've updated my PR with your requested changes. Thanks again for the review!

brikis98 reviewed (approved):
LGTM, thank you! I'll kick off tests now.

brikis98 commented:
Tests passed! Merging now.

brikis98 commented:
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.29.4

thehunt33r commented:
Thank you so much @brikis98!

I'll be waiting for the binaries and starting to use it ASAP :D

ct-hbk commented:
@brikis98 @thehunt33r - thank you guys! I am starting to use it ASAP too
